Apache NiFi integrations, record oriented processor

inigoao opened the pull request:
Alpha version to record oriented Processors for reading PLC values

guszylk committed:
first version of Plc4x record reader tested with s7 protocol

guszylk committed:
fix exception with cause, as the original forked code

guszylk committed:
clean pom file

Commit:
integrations/nifi nueva version del método de Plc4xCommon createSchema, ue utiliza el Record Writer. Posible acceso al dato agnostico respecto al protocolo

Commit:
ajustes Plc4xWriter para inducir protocolo desde Plc4xSourceRecordProcessor, añadidos mapping de datatypes para S7 y Modbus (a record de avro), incluido campo con timestamp en el record

guszylk committed:
TODO comment OnStopped phase

guszylk committed:
Fix enum method

guszylk committed:
Ajuste menor fix para forzar reconexión

guszylk committed:
default value to true

guszylk committed:
1.9.2 nifi version

guszylk committed:
Nifi processor test, not finished yet

guszylk committed:
Pool connection as singleton

guszylk committed:
Fix: reconnect force on exceptions

thomas169 committed:
spi/src/system.c: fixed mem leaks, removed unneede allocs, fixed compiler warnings on hurz cast and degloablised hurz. Rewrote the connection string parser to use tokenisation rather than walking pointer. hello_world_s7.c corrected some printfs and allowed cli args. transport_tcp.c fixed compile warning on printf. hello-wolrd-s7 now working

thomas169 committed:
Added a loopback test which involves getting write values working, issue with s7 payload data item offset to resolve

thomas169 committed:
fix to offsetting of var paload and some exra fetures on loopback tests

thomas169 committed:
revert changes not needed in autogenereted code

thomas169 committed:
Merge branch 'develop' of https://github.com/apache/plc4x into develop

thomas169 committed:
loopback working with intergers

thomas169 committed:
fix for 32bit int writes, now works too

thomas169 committed:
all data types write and arrays

thomas169 committed:
write works with gt 1 byte list items

Commit:
Fork updated with original repo (upstream https://github.com/apache/plc4x.git). Tested the Plc4xSourceRecordProcessor with current PLC4X version (0.9.0-SNAPSHOT) and a S7-1200, working fine. Included information about usage on the README.md in the plc4j/integrations/apache-nifi module

AlvSel committed:
README typos

thomas169 committed:
added fcns to make data item arrays into data.c/.h, added error check to write responces in sm_write.c, correct instalisation of request values in packets.c, naming consistency in sm_read.c.

thomas169 committed:
check read responce items

thomas169 committed:
misc commit

thomas169 committed:
Merge branch 'develop' of https://github.com/apache/plc4x into develop

thomas169 committed:
read returns results to api

inigoao committed:
README markdown correction

ottobackwards commented:
Can you squash or rebase this into a single commit to start?

ottobackwards commented on plc4j/integrations/apache-nifi/README.md:
```diff
@@ -0,0 +1,55 @@
+
+This processor is <ins>record oriented</ins>, formatting output flowfile content using a Record Writer. An example of operation is included on the following path:
+*./test-nifi-template/NIFI-PLC4XIntegration-record-example-1.12.xml*. This file is a Nifi Template that could be directly imported from the NiFi UI to test the operation.
+
```
Maybe a link to the Nifi Documentation

inigoao replied:
done, added link to Apache NiFi RecordPath Guide

ottobackwards commented on plc4j/integrations/apache-nifi/README.md:
```diff
@@ -0,0 +1,55 @@
+*./test-nifi-template/NIFI-PLC4XIntegration-record-example-1.12.xml*. This file is a Nifi Template that could be directly imported from the NiFi UI to test the operation.
+
+The Plc4xSourceRecord Processor can be configured using the following **properties**:
+
```
This is an example where a base + protocol specific derived processors would make thing better for the user, the documentation can be specific to the protocol.

ottobackwards commented on plc4j/integrations/apache-nifi/README.md:
```diff
@@ -0,0 +1,55 @@
+- *PLC resource address String:* *var1=%DB1:DBX0.0:BOOL;var2=%DB1:DBX0.1:BOOL;var3=%DB1:DBB01:BYTE;var4=%DB1:DBW02:WORD;var5=%DB1:DBW04:INT*
+- *Record Writer:* *PLC4x Embedded - AvroRecordSetWriter*
+- *Force Reconnect every request:* *false*
+
```
So the discussion of the schema is important here. 
Here is how the schema is generated by default.
This is the format of the data, if you chose to write your own schema etc.

Would you want to _explicitly_ recommend inheriting the schema?

ottobackwards commented on plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/main/java/org/apache/plc4x/nifi/BasePlc4xProcessor.java:
```diff
@@ -30,14 +30,14 @@ Licensed to the Apache Software Foundation (ASF) under one
 
 public abstract class BasePlc4xProcessor extends AbstractProcessor {
 
```
So, maybe we don't want to use PLC here.  Just Connection String and resource address.  Not everything will be a PLC technically, and if derived processors have to use these properties, it would really stick out.
@chrisdutz thoughts?

ottobackwards replied:
Also, what if we want custom per derived validation?
If the connection strings are common to all drivers, and can all be validated the same, this is OK, if not then this needs to be per processor

Same for the address

ottobackwards replied:
I think we would want to support Expression Language here,  with parameter support.  What if someone has a TEST lab and a Production setup, with different connection strings and or addresses?

They may want to write a flow, and test it and deploy it with the nifi registry and just replace the parameters to the prod devices

ottobackwards commented on plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/main/java/org/apache/plc4x/nifi/BasePlc4xProcessor.java:
```diff
@@ -65,9 +65,14 @@ Licensed to the Apache Software Foundation (ASF) under one
+        this.relationships = new HashSet<>(Arrays.asList(REL_SUCCESS, REL_FAILURE));
     }
 
+    
```
These should be removed.  All access to parameters should be through the context

ottobackwards commented on plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/main/java/org/apache/plc4x/nifi/Plc4xSourceRecordProcessor.java:
```diff
@@ -0,0 +1,269 @@
+        try {
+        	Boolean force =  context.getProperty(FORCE_RECONNECT).asBoolean();
+        	pool = new PooledPlcDriverManager();
+        	if(!force) {
```
access as properties

ottobackwards commented on plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/main/java/org/apache/plc4x/nifi/Plc4xSourceRecordProcessor.java:
```diff
@@ -0,0 +1,269 @@
+	@Override
+	public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+		FlowFile fileToProcess = null;
+		//TODO: In the future the processor will be configurable to get the address and the connection from incoming flowfile
```
This processor is set to INPUT FORBIDDEN, there should never be incoming connection

ottobackwards commented on plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/main/java/org/apache/plc4x/nifi/Plc4xSourceRecordProcessor.java:
```diff
@@ -0,0 +1,269 @@
+				// Prepare the request.
+				if (!connection.getMetadata().canRead()) {
+					throw new ProcessException("Reading not supported by connection");
+				}
```
I'm confused about the input file here

ottobackwards commented on plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/main/java/org/apache/plc4x/nifi/Plc4xSourceRecordProcessor.java:
```diff
@@ -0,0 +1,269 @@
+						session.commit();
+						resultSetFlowFiles.clear();
+					}
+				} catch (Exception e) {
```
English please

ottobackwards replied:
This logic around the try catches seems involved, perhaps a comment explaining what you are doing and why would help maintainability

ottobackwards commented on plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/main/java/org/apache/plc4x/nifi/record/Plc4xReadResponseRecordSet.java:
```diff
@@ -0,0 +1,172 @@
+        }
+        return value;
+    }
+
```
As stated on the list, I think creating the schema once, when the first data is retrieved *from* the data would be best

ottobackwards commented on plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/test/java/org/apache/plc4x/processors/plc4x4nifi/SandBox.java:
```diff
@@ -0,0 +1,47 @@
+package org.apache.plc4x.processors.plc4x4nifi;
+
```
what is this?

ottobackwards reviewed (changes requested):
initial comments

ottobackwards commented:
Also, these processors should have additionalDetails.html  so that there are the extra help files ( with the information that you have in the readme ) for each protocol processor

sruehl committed:
plc4j: fixed inconsistency between xml read/write on booleans

sruehl committed:
plc4j: fixed bug on xml read buffer where newlines broke the number parsing

sruehl committed:
plc4go: added support for driver test suite migration

+ model-template.go.ftlh: removed obsolete enum iteration when enum has content.
+ DriverTestRunner.go: added protocolName and outputFlavor and refactored message parsing to a map based approach with fallback to old xml serialization

sruehl committed:
plc4x: added protocolName and output flavor to driver-testsuite.xsd

sruehl committed:
plc4j: fixed enum rendering

+ Enum should rendered in own context with a uppercase name

sruehl committed:
plc4go: fixed error on push/pop/pull/close not being handled

sruehl committed:
plc4j: refactored test-utils

+ ParserSerializerTestsuiteRunner and DriverTestsuiteRunner have now the ability to convert from the old jackson based xml serialization to the new mpsec serializer based approach
+ split up DriverTestsuiteRunner into multiple classes
+ fixed parser testsuites using the old reduced format

sruehl committed:
plc4x: migrated DriverTestsuite.xml* to new mspec serializer based format.

sruehl committed:
plc4go: fixed enum rendering on indirect enums references

+ fixed issue in TestUtils.go where legit errors got eaten

sruehl committed:
plc4j: fixed issue where we get an superfluous newline in xml

+ prevent WriteBufferXmlBased from writing a newline at the last element
+ fixed TestCasePatcher adding too much newlines to the indented string

sruehl committed:
plc4j: addon for 9b72779b

(Intellij seems to have issues with tracking git state lately)

ottlukas committed:
Update adopters.adoc

ottlukas committed:
Update adopters.adoc

sruehl committed:
plc4go: deprecated old xml/box methods

sruehl committed:
plc4go: initial string implementation

sruehl committed:
plc4go: initial string implementation (part 2)

Commit:
added link to Nifi Documentation in README

inigoao commented:
thank you for all your comments @ottobackwards , i will review them and try to answer and fix all issues.

Commit:
Plc4xSourceRecordProcessor comments translate to english

sruehl committed:
plc4go: improved performance of asciiBox by using strings.Builder

+ Implemented box.Height() method to calculate the height of a box
+ Use string builder whenever possible
+ Implemented benchmark test for box.BoxString (before: 1304 900009 ns/op, after: 4422 243379 ns/op)
+ Implemented benchmark test for box.BoxSideBySide (before: 234 4607450 ns/op, after: 3070	370681 ns/op)
+ Implemented benchmark test for box.expandBox (before: 152 7341252 ns/op, after: 1150	1115318 ns/op)

chrisdutz committed:
Enabled the scan for the full KNX address range

chrisdutz committed:
Updated the KNX discovery example to also allow providing an explicit gateway IP/Hostname

hutcheb committed:
Remove merge button when merging PR's, this forces the use of squash (#238)

sruehl committed:
plc4go: enabled knxnetip parser test

hongjinlin committed:
plc4go: fixed passing parameters incorrectly, resulting in a null pointer (#243)

sruehl committed:
plc4go: implemented write support for SingleItemRequestInterceptor.go

sruehl committed:
plc4go: small cleanup

sruehl committed:
plc4go: added the possibility to get the name and change the name of a asciibox

sruehl committed:
plc4go: fixed a bug in unwrap of AsciiBox where boxes within boxes got wrongfully trimmed

sruehl committed:
plc4go: added an option to WriteBufferBoxBased.go to print boxes in a compact form

sruehl committed:
plc4go: added an option to WriteBufferBoxBased.go to omit empty boxes in a compact form

chrisdutz committed:
Merge branch 'develop' of github.com:apache/plc4x into thomas169_develop

chrisdutz committed:
- Adjusted the PR to work in the maven build on a system without unity installed
- Fixed some problems in the driver_modbus_encode_decode.c function for parsing the address strings (Wonder why this didn't fail on Windows or Mac)

chrisdutz committed:
- Added new KNX Manufacturer

chrisdutz committed:
- Added a BYTE simple type

chrisdutz committed:
- Explicitly excluded H50 from the node-list

sruehl committed:
plc4x: added byte support to readers and writers

+ json and xml based writers treat byte now as hex

sruehl committed:
plc4go: added byte array support to readers and writers

sruehl committed:
plc4go: fixed issue where hex mutated the input array

sruehl committed:
eip/simulated: fixed wrong use of uint

+ simulated added a dummy type to prevent code generation issues in go

sruehl committed:
plc4go: fixed several issues in code generation

+ fixed checksum code
+ partial fixed manual array code
+ disabled broken code
+ prevent broken code on - in protocolname

sruehl committed:
plc4go: added abeth/df1/eip/firmata/simulated

+ Adding these now should help to identify code-generation issues faster

sruehl committed:
plc4j: added byte array support to readers and writers

sruehl committed:
fix build

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

sruehl committed:
plc4x: removed old rendering methods

+ plc4j: removed Jackson annotations from the generated classes to encourage the use of XmlBased Reader/Writers
+ plc4go: removed xml and box rendering methods (string method temporary not available)
+ renamed io to read/write-Buffer to not overlap io imports in golang

sruehl committed:
plc4go: cleanup xml helper code

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

sruehl committed:
plc4j: removed legacy xml serialisation from test-utils

sruehl committed:
plc4x: removed traces from unnecessary java class names from Test suites

sruehl committed:
plc4go: removed un/marshal methods from enums

chrisdutz committed:
PLC4X-298 - [S7] When writing REAL values the Write operation fails with an internal error

chrisdutz committed:
PLC4X-298 - [S7] When writing REAL values the Write operation fails with an internal error

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
nifi-record-oriented processor new version, deduce schema from response

guszylk committed:
fix readme

guszylk committed:
back to original readme

guszylk committed:
FIX the problem with protocols into test, is necessary to include driver explicitly into the pom, with test scope

guszylk committed:
fix: static field, dependencies etc..

guszylk committed:
fix: documentation dependency

guszylk committed:
mvn dependencies analyzer ignore nifi-standard-nar and add as not only provided dependency

guszylk committed:
clean sanbox class from test package

Commit:
updated Plc4xSourceRecordProcessor onScheduled to get address map string during OnTrigger method

guszylk committed:
FIX addressmap with expression language

inigoao committed:
Merge pull request #1 from zylklab/nifi-record-feature

Nifi record feature

inigoao commented:
closing this PR, will upload another one on a dedicated branch 'feature/nifi-record'
